pkg/utils/queue/retry_queue.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/utils/clock"
+
+	"sigs.k8s.io/kwok/pkg/utils/backoff"
+)
+
+// RetryQueue abstracts the behaviour of a queue for retry purpose.
+// Compared to a general Queue type, the meaning of the methods of
+// a retry queue is re-defined.
+// The caller calls Get to peek an item to retry, but the item should
+// not be removed from the underlying state unless the
+// caller explicitly calls the Remove method. The caller MUST also
+// explicitly re-Add the item to the queue if a retry fails or Remove it
+// if it succeeds.
+type RetryQueue[T comparable] interface {
```

Comment:
It doesn't feel like a queue anymore. Wouldn't it be more appropriate to put it in "sigs.k8s.io/kwok/pkg/utils/wait"?